text/0000-consensus-api.md

```diff
@@ -0,0 +1,320 @@
+
+**R5 - Read/Write Access to Global State**
+
+PoET depends on tracking authorized block publishers in global state and
+therefore requires read access to a PoET specific namespace.
+
+**R6 - Reusability**
+
+The Consensus Engine interface and implementations of the interface should be
+reusable in other applications. Specifically, the interface should support
+non-blockchain applications.
+
+## Data Structures
+
+**Blocks**
+
+In order to support R6, "block" is defined to mean the following data structure
+within the context of the consensus engine. When used in a blockchain context,
+this block definition is equivalent to those parts of the block that are
+relevant to consensus. When used outside the blockchain context, the
+interpretation of this data structure is context dependent.
+
+    Block {
+        Id: string
+        Previous Id: string
+        Index: unsigned integer
+        Consensus: bytes
+    }
+
+**Consensus Messages**
+
+In order to support arbitrary communication between nodes, a generic message
+type is defined with a payload that is opaque to the validator. The following
+data structure is used for this:
+
+    ConsensusMessage {
+        MessageType: string
+        Payload: bytes
+    }
+
+Payload is the opaque payload to send to other nodes. The interpretation of
+MessageType is left to the Consensus Engine implementation.
+
+## API
+
+The Consensus Engine API below is presented as a set of methods in order to
+simplify their presentation and explanation. Each method is implemented with a
+pair of (Request, Response) messages and the Consensus Proxy is responsible for
+handling this transformation within the Validator.
+
+The methods presented below are only meant to describe roughly what the
+high-level interactions between the Consensus Engine and the Validator will be.
+The names, parameters, and exact behavior of these methods are subject to
+change based on additional constraints and requirements discovered during
+implementation.
+
+The methods defined by the Consensus Engine interface are split into the
+following groups:
+
+1. P2P Messaging
+2. Event Handlers
+3. Block Creation
+4. Block Management
+5. Queries
+6. Engine
+
+### P2P Messaging Methods
+The following methods are provided to Consensus Engines for sending consensus
+messages to other nodes on the network. These methods support R1.
+
+| Method | Description |
+| --- | --- |
+| SendTo(peer, message) | Send a consensus message to a specific, connected peer |
+| Broadcast(message) | Broadcast a message to all peers |
+
+### Event Handlers
+The following methods will be implemented by Consensus Engines. They are used
+to update the Consensus Engine on external events. These methods support R1-R3.
+
+| Method | Description |
+| --- | --- |
+| OnMessageReceived(message) | Called when a new consensus message is received |
+| OnNewBlockReceived(block) | Called when a new block is received and validated |
```

Comment:
Does "and validated" imply that validator will validate transactions before consensus considers whether the block is valid or has precedence?